repo/format/format_manager.go

```diff
@@ -0,0 +1,466 @@
+package format
+
+import (
+	"context"
+	"crypto/rand"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/internal/feature"
+	"github.com/kopia/kopia/internal/gather"
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/encryption"
+	"github.com/kopia/kopia/repo/hashing"
+	"github.com/kopia/kopia/repo/logging"
+)
+
+var log = logging.Module("kopia/repo/format")
+
+// UniqueIDLength is the length of random unique ID of each repository.
+const UniqueIDLength = 32
+
+// Manager manages the contents of `kopia.repository` and `kopia.blobcfg`.
+type Manager struct {
+	//nolint:containedctx
+	ctx           context.Context // +checklocksignore
+	blobs         blob.Storage    // +checklocksignore
+	validDuration time.Duration   // +checklocksignore
+	password      string          // +checklocksignore
+	cache         BlobCache       // +checklocksignore
+
+	timeNow func() time.Time // +checklocksignore
+
+	// all the stuff protected by a mutex is valid until `validUntil`
+	mu sync.RWMutex
+	// +checklocks:mu
+	formatEncryptionKey []byte
+	// +checklocks:mu
+	j *KopiaRepositoryJSON
+	// +checklocks:mu
+	repoConfig *RepositoryConfig
+	// +checklocks:mu
+	blobCfgBlob BlobStorageConfiguration
+	// +checklocks:mu
+	current Provider
+	// +checklocks:mu
+	validUntil time.Time
+	// +checklocks:mu
+	loadedTime time.Time
+	// +checklocks:mu
+	refreshCounter int
+}
+
+func (m *Manager) getFormat() Provider {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.current
+}
+
+func (m *Manager) getOrRefreshFormat() (Provider, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if err := m.maybeRefreshLocked(); err != nil {
+		return nil, err
+	}
+
+	return m.current, nil
+}
+
+// +checklocks:m.mu
+func (m *Manager) maybeRefreshLocked() error {
+	if n := m.timeNow(); !n.Before(m.validUntil) {
```

Comment:
Nit: early return is possible.  Return `nil` when `n := m.timeNow(); n.Before(m.validUntil)`